### PR TITLE
Explore: Add feature tracking events

### LINKED ISSUE #54082: Implementation of Tracking Items

### PR DESCRIPTION
**Pull request by L-M-K-B**
Fixes #54082 

- tracking events in `QueryEditorRow.tsx`: tracking is only available for the component in Explore, other use cases were excluded

### CODE REVIEW

File: public/app/features/query/components/QueryEditorRow.tsx [hunk 1]
```diff
@@ -273,18 +280,32 @@ export class QueryEditorRow<TQuery extends DataQuery> extends PureComponent<Prop
...
-    const { query } = this.props;
-    this.props.onChange({ ...query, hide: !query.hide });
-    this.props.onRunQuery();
+    const { query, onChange, onRunQuery, trackActions } = this.props;
+    onChange({ ...query, hide: !query.hide });
+    onRunQuery();
+
+    if (trackActions) {
+      reportInteraction(trackActions.disableEnableQuery);
+    }
```
Comment (Elfo404) [inline public/app/features/query/components/QueryEditorRow.tsx:308]: i'm wondering wether this belongs here. This component shouldn't care at all about tracking, and maybe we should bubble up all those events so that they can be handled on the consumer side, which can then add tracking.

File: public/app/features/explore/ExploreTimeControls.tsx [hunk 2]
```diff
@@ -44,6 +45,12 @@ export class ExploreTimeControls extends Component<Props> {
       from: adjustedFrom,
       to: adjustedTo,
     });
+
+    if (typeof adjustedFrom === 'string' && adjustedFrom.startsWith('now-')) {
```
[the remark this answers is not in the document]
Comment (L-M-K-B) [reply public/app/features/explore/ExploreTimeControls.tsx:49]: > if the user changes the range manually like below it will be tracked with the same event
I know and that's not what we intend to measure. The reason why I implemented it anyway is that it is the closest we can get to what we are actually interested in. With tracking all range selections I am a bit worried about the amount of data we would get and don't use. So what about tracking only the time ranges with timeRange.raw.to = "now"?

File: public/app/features/explore/ExploreToolbar.tsx [hunk 3]
```diff
@@ -38,7 +38,12 @@ type Props = OwnProps & ConnectedProps<typeof connector>;
 
 class UnConnectedExploreToolbar extends PureComponent<Props> {
   onChangeDatasource = async (dsSettings: DataSourceInstanceSettings) => {
-    this.props.changeDatasource(this.props.exploreId, dsSettings.uid, { importQueries: true });
+    const { changeDatasource, datasourceRef } = this.props;
+    changeDatasource(this.props.exploreId, dsSettings.uid, { importQueries: true });
+    reportInteraction('grafana_explore_ds_picker_ds_change', {
```
[the remark this answers is not in the document]
Comment (L-M-K-B) [reply public/app/features/explore/ExploreToolbar.tsx:43]: That's true.

File: public/app/features/explore/QueryRows.tsx [hunk 4]
```diff
@@ -85,6 +91,7 @@ export const QueryRows = ({ exploreId }: Props) => {
       app={CoreApp.Explore}
       history={history}
       eventBus={eventBridge}
+      trackActions={trackActions}
```
[the remark this answers is not in the document]
Comment (L-M-K-B) [reply public/app/features/explore/QueryRows.tsx:94]: I agree.

File: public/app/features/query/components/QueryEditorRow.tsx (hunk 1 above)
Comment (Elfo404) [reply public/app/features/query/components/QueryEditorRow.tsx]: can you please explain how adding a standard`onQueryCopied` or something similar (which this component already partially has, and is widely used) is more confusing than a trackingContext property that when set makes the component track interaction with hardcoded names and values?

File: public/app/features/explore/QueryRows.tsx [hunk 5]
```diff
@@ -44,6 +44,12 @@ export const QueryRows = ({ exploreId }: Props) => {
   const history = useSelector(getHistory);
   const eventBridge = useSelector(getEventBridge);
 
+  const trackActions = {
+    duplicateQuery: 'grafana_explore_duplicate_query',
+    disableEnableQuery: 'grafana_explore_disable_enable_query',
```
[the remark this answers is not in the document]
Comment (L-M-K-B) [reply public/app/features/explore/QueryRows.tsx]: Added it.

File: public/app/features/explore/ExploreTimeControls.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (ifrost) [reply public/app/features/explore/ExploreTimeControls.tsx:49]: Not sure if we need to be worried about the amount of data. In my opinion, if we can't get the correct data it's better to track something else (but correctly) or not to track. Having tracking data that may not represent what we're supposed to track may cause making incorrect decisions based on that data.

However, if we need to get "close enough" to quick time ranges and track it this way I would make sure it's clear what we're tracking:
* `grafana_explore_time_picker_time_range` -> we need to change the name of the event because it's not tracking all range selections. It could be for example `grafana_explore_time_picker_relative_to_time_range_selected`
* we need to make a comment in the code about why we're tracking relative ranges only (i.e. because we want to find out what quick choices are most popular).

File: public/app/features/query/components/QueryEditorRow.tsx (hunk 1 above)
Comment (L-M-K-B) [reply public/app/features/query/components/QueryEditorRow.tsx]: Just to get it right let me rephrase what I understood from your comments. Please correct me if I got it wrong:
- create 3 functions for tracking in _public/app/features/explore/**QueryRows.tsx**_ for `onQueryCopied`, `onQueryRemoved` and `onQueryEnabled`
- pass them through _public/app/features/query/components/**QueryEditorRows.tsx**_ to _public/app/features/query/components/**QueryEditorRow.tsx**_ so they can get called when one of the 3 buttons is clicked
- So every time someone wants to implement tracking for their use case in _public/app/features/query/components/**QueryEditorRow.tsx**_  they have to create their own functions because we would like to keep the component in grafana-ui as general as possible

This sounds unnecessary complex and there is plenty of code that could not be reused. I am not questioning the practice of passing through callbacks but in this case it seems too much.
Currently the component in Explore is passing a simple string to the ui component. It is obvious where it is being used and for which use case. To make it even more obvious I could leave a comment what is expected in the string.

I implemented the tracking in a component that is widely used because there are plenty of teams working on tracking and they can reuse it.

What do you think?

File: public/app/features/explore/ExploreTimeControls.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (L-M-K-B) [reply public/app/features/explore/ExploreTimeControls.tsx:49]: I agree that tracking the wrong things could lead to incorrect conclusions. If we don't have an issue with too much data let's track all time ranges. So we are going to have all the data and we can choose what we are interested in afterwards (e.g. in the dashboard). 
I will adjust the tracking there.

File: public/app/features/query/components/QueryEditorRow.tsx (hunk 1 above)
Comment (Elfo404) [reply public/app/features/query/components/QueryEditorRow.tsx]: i think this component responsibility does not include tracking, and it should be implemented separately by its consumers.
it may seem obvious to you how this component works, but it may not be to everyone. what you are proposing here is not done anywhere else.

Comment (ifrost) [reply public/app/features/query/components/QueryEditorRow.tsx]: There's already tracking in `QueryEditorRows`. `QueryEditorRow` is just an internal component used by "rows" container. That means it will be consistent with the current approach to add tracking to `QueryRow` (not sure if it's the best way but at least we should be consistent so we can change it later more easily). 

The concerns you both have are valid and it all depends on how we treat `QueryEditorRow(s)` component. If we ever plan to make it generic as `grafana-ui` components it cannot have a dependency to `reportInteractions` (it's part of the runtime and it would make it not reusable), but that would also require moving out other events. If the purpose of this component is to be used only in core Grafana we should probably:

* make names consistent - e.g. we have `query_row_reorder_ended` and `grafana_queryRow_remove_query` 
* `trackingContext` should also be added to other dragging events, and possibly to all consumers of QueryEditorRow
* unless we want to track these actions regardless of "context", then we can add tracking without it.

At the moment what is probably confusing is that we limit some interaction tracking to Explore only.

Comment (ivanortegaalba) [reply public/app/features/query/components/QueryEditorRow.tsx]: IMO, we should be careful when adding tracking to reusable components. They can break other people's metrics if you use them as part of a funnel since you can start getting more traffic that doesn't belong to your experience. 

I suggest not reporting interaction from the component itself but doing it via callback from the upper level that is used only once and it belongs only to a product.

File: public/app/features/explore/ExploreTimeControls.tsx [hunk 6]
```diff
@@ -44,6 +45,11 @@ export class ExploreTimeControls extends Component<Props> {
       from: adjustedFrom,
       to: adjustedTo,
     });
+
+    reportInteraction('grafana_explore_time_picker_time_range', {
```
Comment (ivanortegaalba) [inline public/app/features/explore/ExploreTimeControls.tsx:49]:
```suggestion
    reportInteraction('grafana_explore_time_picker_time_range_changed', {
```

File: public/app/features/query/components/QueryEditorRow.tsx (hunk 1 above)
Comment (ivanortegaalba) [reply public/app/features/query/components/QueryEditorRow.tsx]: About naming, using grafana as the namespace is too vague, I'd suggest using: `grafana_explore` as a prefix to identify easily all the product events as you do in the other events. WDYT?

**Review by ivanortegaalba — COMMENTED** [review]
I added some small suggestions, thanks  @L-M-K-B 💯

File: public/app/features/query/components/QueryEditorRow.tsx [hunk 7]
```diff
@@ -57,6 +57,7 @@ interface Props<TQuery extends DataQuery> {
   history?: Array<HistoryItem<TQuery>>;
   eventBus?: EventBusExtended;
   alerting?: boolean;
+  onTriggerTracking?: (action: string, queryStatus?: boolean | undefined) => void;
```
Comment (ifrost) [inline public/app/features/query/components/QueryEditorRow.tsx:60]: I think the method shouldn't be called `onTriggerTracking` as it can be used to handle anything (maybe something like `onEvent` would be better?). Also the optional `queryStatus` param is confusing because it's passed only with `enableDisable` so it's harder to understand just reading the API. @Elfo404, @ivanortegaalba please correct me if I'm wrong by I think what you  meant was to have separate handlers for each event:

```typescript
onQueryRemoved: () => void;
onQueryCopied: () => void;
onQueryToggled: (shown: boolean): void;
```

Comment (L-M-K-B) [reply public/app/features/query/components/QueryEditorRow.tsx]: Done.

File: public/app/features/query/components/QueryEditorRow.tsx [hunk 8]
```diff
@@ -57,6 +57,9 @@ interface Props<TQuery extends DataQuery> {
   history?: Array<HistoryItem<TQuery>>;
   eventBus?: EventBusExtended;
   alerting?: boolean;
+  onQueryCopied?: () => void;
+  onQueryRemoved?: () => void;
+  onQueryToggled?: (queryStatus?: boolean | undefined) => void;
```
Comment (ifrost) [inline public/app/features/query/components/QueryEditorRow.tsx:62]: `queryStatus` is always passed to this callback so we can skip `| undefined`

```suggestion
  onQueryToggled?: (queryStatus: boolean) => void;
```

File: public/app/features/query/components/QueryEditorRows.tsx [hunk 9]
```diff
@@ -31,7 +31,9 @@ interface Props {
   app?: CoreApp;
   history?: Array<HistoryItem<DataQuery>>;
   eventBus?: EventBusExtended;
-
+  onQueryCopied?: () => void;
+  onQueryRemoved?: () => void;
+  onQueryToggled?: (queryStatus?: boolean | undefined) => void;
```
Comment (ifrost) [inline public/app/features/query/components/QueryEditorRows.tsx:36]:
```suggestion
  onQueryToggled?: (queryStatus: boolean) => void;
```

**Review by ifrost — APPROVED** [review]
Great stuff 🎉

File: public/app/features/query/components/QueryEditorRows.tsx (hunk 9 above)
Comment (L-M-K-B) [reply public/app/features/query/components/QueryEditorRows.tsx:36]: See above.

File: public/app/features/query/components/QueryEditorRow.tsx (hunk 8 above)
Comment (ifrost) [reply public/app/features/query/components/QueryEditorRow.tsx:62]: It could have been casted to a boolean.
